### PR TITLE
DISTMYSQL-36: Create pipeline for releasing distributions

### PR DESCRIPTION
**Pull request by Sudokamikaze**

### CODE REVIEW

File: ps-pxc-dist/apt_release.sh [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+for $dir in $DIRECTORIES; do
+    #
+    set -e
+    echo "<*> reprepro binary is $(which reprepro)"
+    pushd /srv/UPLOAD/$dir/binary/debian
+    echo "Looking for Debian build directories..."
+    if [[ ${COMPONENT} != RELEASE ]]; then
+        CODENAMES=$(ls -1)
+    else
+        CODENAMES=$(ls -1 | egrep -v "cosmic|disco" )
```
Comment (EvgeniyPatlan) [inline ps-pxc-dist/apt_release.sh:45]: we don't need it any more

File: ps-pxc-dist/Jenkinsfile.groovy [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        stage('push-to-deb-repository') {
+            steps {
+                sh '''
+                    wget https://raw.githubusercontent.com/Sudokamikaze/jenkins-pipelines/DISTMYSQL-36/ps-pxc-dist/apt_release.sh
+                    bash -xe apt_release.sh
+                '''
+            }
+        }
+    }
+
```
Comment (EvgeniyPatlan) [inline ps-pxc-dist/Jenkinsfile.groovy:98]: please add sync repos stage and clean_cdn_cache

File: ps-pxc-dist/apt_release.sh (hunk 1 above)
Comment (Sudokamikaze) [reply ps-pxc-dist/apt_release.sh:45]: Will remove

File: ps-pxc-dist/Jenkinsfile.groovy (hunk 2 above)
Comment (Sudokamikaze) [reply ps-pxc-dist/Jenkinsfile.groovy:154]: Added few seconds ago :)

File: ps-pxc-dist/Jenkinsfile.groovy [hunk 3]
```diff
@@ -0,0 +1,233 @@
+pipeline {
+    agent {
+        label 'min-buster-x64'
```
Comment (EvgeniyPatlan) [inline ps-pxc-dist/Jenkinsfile.groovy:3]: better to use source_builder

File: ps-pxc-dist/Jenkinsfile.groovy [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+                    rsync ${RSYNC_TRANSFER_OPTS} --exclude=*.sh --exclude=*.bak /srv/repo-copy/${REPO}/* 10.10.9.209:/www/repo.percona.com/htdocs/${REPO}/
+                '''
+            }
+        }
+        stage('clear-cdn-cache') {
+            when {
+                expression { params.SKIP_CDN == false }
+            }
+            steps {
+                sh '''
```
Comment (EvgeniyPatlan) [inline ps-pxc-dist/Jenkinsfile.groovy:156]: https://github.com/Percona-Lab/jenkins-pipelines/blob/master/vars/sync2Prod.groovy#L66 please look howit is done for PMM team

File: ps-pxc-dist/rpm_release.sh [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+for REPOPATH in $REPOPATH_TMP; do
+    for $dir in $DIRECTORIES; do
+        #
+        cd /srv/UPLOAD/$dir
+        #
+        # getting the list of RH systems
+        RHVERS=$(ls -1 binary/redhat)
+        #
+        # source processing
+        SRCRPM=$(find source/redhat -name '*.src.rpm')
```
Comment (EvgeniyPatlan) [inline ps-pxc-dist/rpm_release.sh:42]: please add check if source directory is available.
There are some cases once we don't have it. So we should push packages without source packages

File: ps-pxc-dist/apt_release.sh [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        #
+        if [[ ${REMOVE_LOCKFILE} = true ]]; then
+            echo "Removing lock file as requested..."
+            sudo rm -vf  ${REPOPATH}/db/lockfile
+        fi
+        #
+        if [[ ${REPOCOMP} = release ]]; then
+            export REPOCOMP=main
+        # pushing sources
+            cd /srv/UPLOAD/$dir/source/debian
```
Comment (EvgeniyPatlan) [inline ps-pxc-dist/apt_release.sh:68]: please add check if source directory is available.
There are some cases once we don't have it. So we should push packages without source packages

File: ps-pxc-dist/apt_release.sh [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+if [[ -z "${DIRECTORIES}" ]]; then
+    echo "No PATH specified..."
+    exit 1
+fi
+#
+REPOCOMP=$(echo "${COMPONENT}" | tr '[:upper:]' '[:lower:]')
+REPOPUSH_ARGS=""
+
+#if [[ ${REMOVE_BEFORE_PUSH} = true ]] || [[ ${COMPONENT} != RELEASE ]]; then
+if [[ ${REMOVE_BEFORE_PUSH} = true ]]; then
```
Comment (EvgeniyPatlan) [inline ps-pxc-dist/apt_release.sh:21]: please return [[ ${COMPONENT} != RELEASE ]]

File: ps-pxc-dist/apt_release.sh [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+    REPOPUSH_ARGS+=" --remove-package"
+fi
+#fi
+#
+#
+if [[ ${REPOSITORY} == "PDPXC" ]]; then
+  REPOPATH_TMP="/srv/repo-copy/pdpxc-${REPOSITORY_VERSION}/apt"
+  if [ $REPOSITORY_VERSION_MAJOR == true ]; then
```
Comment (EvgeniyPatlan) [inline ps-pxc-dist/apt_release.sh:29]:
```suggestion
  if [[ "${REPOSITORY_VERSION_MAJOR}" == "true" ]]; then
```

File: ps-pxc-dist/apt_release.sh [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+REPOPUSH_ARGS=""
+
+#if [[ ${REMOVE_BEFORE_PUSH} = true ]] || [[ ${COMPONENT} != RELEASE ]]; then
+if [[ ${REMOVE_BEFORE_PUSH} = true ]]; then
+    REPOPUSH_ARGS+=" --remove-package"
+fi
+#fi
+#
+#
+if [[ ${REPOSITORY} == "PDPXC" ]]; then
```
Comment (EvgeniyPatlan) [inline ps-pxc-dist/apt_release.sh:27]:
```suggestion
if [[ "${REPOSITORY}" == "PDPXC" ]]; then
```

File: ps-pxc-dist/apt_release.sh [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+  REPOPATH_TMP="/srv/repo-copy/pdpxc-${REPOSITORY_VERSION}/apt"
+  if [ $REPOSITORY_VERSION_MAJOR == true ]; then
+    REPOPATH_TMP+=" /srv/repo-copy/pdpxc-$(echo ${REPOSITORY_VERSION} | awk -F"." 'BEGIN { OFS = "." }{ print $1, $2}' | sed 's/\.$//')/apt"
+  fi
+  export PATH="/usr/local/reprepro5/bin:${PATH}"
+fi
+
+if [[ ${REPOSITORY} == "PDPS" ]]; then
```
Comment (EvgeniyPatlan) [inline ps-pxc-dist/apt_release.sh:35]:
```suggestion
if [[ "${REPOSITORY}" == "PDPS" ]]; then
```

File: ps-pxc-dist/apt_release.sh [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+  REPOPATH_TMP="/srv/repo-copy/pdps-${REPOSITORY_VERSION}/apt"
+  if [ $REPOSITORY_VERSION_MAJOR == true ]; then
```
Comment (EvgeniyPatlan) [inline ps-pxc-dist/apt_release.sh:37]:
```suggestion
  if [[ "${REPOSITORY_VERSION_MAJOR}" == "true" ]]; then
```

File: ps-pxc-dist/apt_release.sh [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+fi
+#
+for REPOPATH in $REPOPATH_TMP; do
+    for dir in $DIRECTORIES; do
+        #
+        set -e
+        echo "<*> reprepro binary is $(which reprepro)"
+        pushd /srv/UPLOAD/$dir/binary/debian
+        echo "Looking for Debian build directories..."
+        if [[ ${COMPONENT} != RELEASE ]]; then
```
Comment (EvgeniyPatlan) [inline ps-pxc-dist/apt_release.sh:50]:
```suggestion
        if [[ "${COMPONENT}" != "RELEASE" ]]; then
```

File: ps-pxc-dist/apt_release.sh [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+        fi
+
+        echo Distributions are: ${CODENAMES}
+        popd
+        #
+        #######################################
+        # source pushing, it's a bit specific #
+        #######################################
+        #
+        if [[ ${REMOVE_LOCKFILE} == true ]]; then
```
Comment (EvgeniyPatlan) [inline ps-pxc-dist/apt_release.sh:61]:
```suggestion
        if [[ "${REMOVE_LOCKFILE}" == "true" ]]; then
```

File: ps-pxc-dist/apt_release.sh [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+            echo "Removing lock file as requested..."
+            sudo rm -vf ${REPOPATH}/db/lockfile
+        fi
+        #
+        if [[ ${REPOCOMP} == release ]]; then
```
Comment (EvgeniyPatlan) [inline ps-pxc-dist/apt_release.sh:66]:
```suggestion
        if [[ "${REPOCOMP}" == "release" ]]; then
```
Also shouldn't release be uppercased?

File: ps-pxc-dist/rpm_release.sh [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+    echo "No PATH specified..."
+    exit 1
+fi
+#
+REPOCOMP=$(echo "${COMPONENT}" | tr '[:upper:]' '[:lower:]')
+REPOPATH=${REPOSITORY}
+#
+if [[ ${REPOSITORY} == "PDPXC" ]]; then
+  REPOPATH_TMP="repo-copy/pdpxc-${REPOSITORY_VERSION}/yum"
+  if [ $REPOSITORY_VERSION_MAJOR == true ]; then
```
Comment (EvgeniyPatlan) [inline ps-pxc-dist/rpm_release.sh:22]:
```suggestion
  if [[ $REPOSITORY_VERSION_MAJOR == true ]]; then
```

File: ps-pxc-dist/rpm_release.sh [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+    REPOPATH_TMP+=" repo-copy/pdpxc-$(echo ${REPOSITORY_VERSION} | awk -F"." 'BEGIN { OFS = "." }{ print $1, $2}' | sed 's/\.$//')/yum"
+  fi
+fi
+#
+if [[ ${REPOSITORY} == "PDPS" ]]; then
+  REPOPATH_TMP="repo-copy/pdps-${REPOSITORY_VERSION}/yum"
+  if [ $REPOSITORY_VERSION_MAJOR == true ]; then
```
Comment (EvgeniyPatlan) [inline ps-pxc-dist/rpm_release.sh:29]:
```suggestion
  if [[ $REPOSITORY_VERSION_MAJOR == true ]]; then
```

File: ps-pxc-dist/rpm_release.sh [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+if [[ "${REPOSITORY}" == "PDPXC" ]]; then
+  REPOPATH_TMP="repo-copy/pdpxc-${REPOSITORY_VERSION}/yum"
+  if [ "${REPOSITORY_VERSION_MAJOR}" == "true" ]; then
+    REPOPATH_TMP+=" repo-copy/pdpxc-$(echo ${REPOSITORY_VERSION} | awk -F"." 'BEGIN { OFS = "." }{ print $1, $2}' | sed 's/\.$//')/yum"
+  fi
+fi
+#
+if [[ "${REPOSITORY}" == "PDPS" ]]; then
+  REPOPATH_TMP="repo-copy/pdps-${REPOSITORY_VERSION}/yum"
+  if [ "${REPOSITORY_VERSION_MAJOR}" == "true" ]; then
```
Comment (EvgeniyPatlan) [inline ps-pxc-dist/rpm_release.sh:29]:
```suggestion
  if [[ "${REPOSITORY_VERSION_MAJOR}" == "true" ]]; then
```

**Pull request merged**